src/main/scala/nl.knaw.dans.dd.prepub/DansDataVaultMetadataBlockMapper.scala

```diff
@@ -35,22 +37,55 @@ class DansDataVaultMetadataBlockMapper(pidGeneratorBaseUrl: URI) {
     val fields = ListBuffer[MetadataField]()
     fields.append(PrimitiveSingleValueField("dansDataversePid", workFlowVariables.pid))
     fields.append(PrimitiveSingleValueField("dansDataversePidVersion", s"${ workFlowVariables.majorVersion }.${ workFlowVariables.minorVersion }"))
-
-    // TODO: How find out if a bagId that is found here is carried over from the previous version or a bagId minted by the SWORD service?
-    //       Compare with bagId of previous version if exists
-    //       if no previous version: filled in bagId means SWORD filled it in
-    //       if previous version: new bagId means SWORD filled it in, the same bagId as in previous version means UI created new draft.
-    fields.append(PrimitiveSingleValueField("dansBagId", b.getOrElse(mintBagId())))
+    fields.append(PrimitiveSingleValueField("dansBagId", setBagId(b, workFlowVariables.pid)))
     fields.append(PrimitiveSingleValueField("dansNbn", n.getOrElse(mintUrnNbn())))
     o.foreach(b => fields.append(PrimitiveSingleValueField("dansOtherId", b)))
     ov.foreach(b => fields.append(PrimitiveSingleValueField("dansOtherIdVersion", b)))
     st.foreach(b => fields.append(PrimitiveSingleValueField("dansSwordToken", b)))
     FieldList(fields.toList)
   }
 
+  private def setBagId(bagIdOpt: Option[String], pid: String): String = {
+    bagIdOpt match {
```

Comment:
Use `.map` and `.getOrElse` on Options.

Reply:
I can't seem to make that work properly.
I also think it is more readable this way.

Reply:
Options and its higher order functions are the bread and butter of Scala, so that should not be so hard to implement. If you find it more readible, that is fine, but what is the point of using Scala if we are not going to use the more basic stuff in it. I can understand that the more advanced niceties may be a bridge too far, but this really makes me wonder.